lib/spack/llnl/util/tty/color.py

```diff
@@ -247,23 +250,24 @@ def __call__(self, match):
         return self.escape(string) + colored_text
 
 
-def colorize(string, **kwargs):
+def colorize(string: str, color: Optional[bool] = None, enclose: Optional[bool] = None) -> str:
     """Replace all color expressions in a string with ANSI control codes.
 
     Args:
-        string (str): The string to replace
-
-    Returns:
-        str: The filtered string
-
-    Keyword Arguments:
-        color (bool): If False, output will be plain text without control
+        string: The string to replace
+        color: If False, output will be plain text without control
             codes, for output to non-console devices.
-        enclose (bool): If True, enclose ansi color sequences with
+        enclose: If True, enclose ansi color sequences with
             square brackets to prevent misestimation of terminal width.
     """
-    color = _color_when_value(kwargs.get("color", get_color_when()))
-    string = re.sub(color_re, match_to_ansi(color, kwargs.get("enclose")), string)
+    color_arg = color if color is not None else get_color_when()
+    return _colorize(string, color=color_arg, enclose=enclose)
+
+
+@memoized
+def _colorize(string: str, color: Optional[bool] = None, enclose: Optional[bool] = None) -> str:
```

Comment:
Not immediately a fan of adding memoization to what is supposed to be a stream wrapper. Better ensure this function only gets called where it should.

Reply:
Ok, let's revert this commit and get the other optimization in. For reference:

This is the effect of using `str()` as keys:

![radiuss-1](https://github.com/spack/spack/assets/4199709/4f838925-d47d-435c-a7df-d90b570dc30c)

and this is the effect of avoiding a lot of `re.sub` calls:

![radiuss-2](https://github.com/spack/spack/assets/4199709/8c620eed-8b75-42f4-a61a-7263012a682f)

Reply:
Apparently develop is also faster today, so maybe there's one of the optimization that we did for the DB that mattered.